sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_read.h>
+
+#include <az_digit.h>
+
+#include <math.h>
+
+inline bool az_json_is_white_space(char const c) {
+  switch (c) {
+    case ' ': case '\t': case '\n': case '\r':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(char const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) {
+  return p->stack == 1;
+}
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) {
+  return p->stack & 1;
+}
+
+az_json_state az_json_state_create(az_const_str const buffer) {
+  return (az_json_state){
+    .buffer = buffer,
+    .i = 0,
+    .stack = 1,
+  };
+}
+
+az_result az_json_get_char(az_const_str const buffer, size_t const i, char *const out_char) {
+  if (i == buffer.size) {
+    return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  *out_char = az_const_str_item(buffer, i);
+  return AZ_OK;
+}
+
+az_result az_json_expect_char(az_const_str const buffer, size_t *const p, char const expected) {
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(buffer, *p, &c));
+  if (c != expected) {
+    return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+  }
+  ++*p;
+  return AZ_OK;
+}
+
+az_result az_json_read_white_space(az_const_str const buffer, size_t *const p) {
+  for (; *p != buffer.size; ++*p) {
+    if (!az_json_is_white_space(az_const_str_item(buffer, *p))) {
+      break;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_stack_push(az_json_state *const p_state, az_json_stack const stack) {
+  if (p_state->stack >> AZ_JSON_STACK_SIZE != 0) {
+    return AZ_JSON_ERROR_STACK_OVERFLOW;
+  }
+  p_state->stack = (p_state->stack << 1) | stack;
+  return AZ_OK;
+}
+
+az_result az_json_read_keyword_rest(
```

Comment:
is it necessary to compare each keyword all the time? seems like when you call this you'd have already compared the keywords and you just need to skip over them.

Reply:
I do compare only once,to make sure that this is `true` instead of `trjiosd`.

Reply:
so as you're comparing you would handle the AZ_JSON_ERROR_UNEXPECTED_CHAR?

Reply:
It will return `AZ_JSON_ERROR_UNEXPECTED_CHAR` if the parser couldn't match the string with the keyword.